scripts/mod_loader/modapi/board.lua

```diff
@@ -1,96 +1,691 @@
+BoardClass.IsForest = function(self, loc)
+	Tests.AssertSignature{
+		ret = "bool",
+		func = "IsForest",
+		params = { self, loc },
+		{ "userdata|GameBoard&", "userdata|Point" }
+	}
+	
+	return CUtils.IsTileForest(self, loc)
+end
 
-	return pawnStack
+BoardClass.IsForestFire = function(self, loc)
+	Tests.AssertSignature{
+		ret = "bool",
+		func = "IsForestFire",
+		params = { self, loc },
+		{ "userdata|GameBoard&", "userdata|Point" }
+	}
+	
+	return CUtils.IsForestFire(self, loc)
 end
 
-BoardClass.RestorePawnsToTile = function(self, loc, pawnStack)
-	Tests.AssertEquals("userdata", type(self), "Argument #0")
-	Tests.AssertTypePoint(loc, "Argument #1")
-	Tests.AssertEquals("table", type(pawnStack), "Argument #2")
+BoardClass.IsShield = function(self, loc)
+	Tests.AssertSignature{
+		ret = "bool",
+		func = "IsShield",
+		params = { self, loc },
+		{ "userdata|GameBoard&", "userdata|Point" }
+	}
+	
+	return CUtils.IsTileShield(self, loc)
+end
 
-	while not pawnStack:isEmpty() do
-		local pawn = pawnStack:popLeft()
-		pawn:SetSpace(loc)
+BoardClass.SetShield = function(self, loc, shield, no_animation)
+	Tests.AssertSignature{
+		ret = "void",
+		func = "SetShield",
+		params = { self, loc, shield, no_animation },
+		{ "userdata|GameBoard&", "userdata|Point", "boolean|bool", "boolean|bool" },
+		{ "userdata|GameBoard&", "userdata|Point", "boolean|bool" },
+		{ "userdata|GameBoard&", "userdata|Point" }
+	}
+	
+	-- Shielding empty tiles is possible, but they are not stored in the savegame,
+	-- so that shouldn't be part of the mod loader.
+	local iTerrain = self:GetTerrain(loc)
+	local isShieldableTerrain = iTerrain == TERRAIN_BUILDING or iTerrain == TERRAIN_MOUNTAIN
+	
+	if shield == nil then
+		shield = true
+	end
+	
+	if no_animation == nil then
+		no_animation = false
+	end
+	
+	if isShieldableTerrain then
+		if no_animation then
+			CUtils.SetTileShield(self, loc, shield)
+		else
+			local d = SpaceDamage(loc)
+			d.iShield = shield and 1 or -1
+			self:DamageSpace(d)
+		end
+	elseif self:IsPawnSpace(loc) then
+		self:GetPawn(loc):SetShield(shield, no_animation)
 	end
 end
 
-BoardClass.SetFire = function(self, loc, fire)
-	Tests.AssertEquals("userdata", type(self), "Argument #0")
-	Tests.AssertTypePoint(loc, "Argument #1")
-	Tests.AssertEquals("boolean", type(fire), "Argument #2")
+BoardClass.SetWater = function(self, loc, water, sink)
+	Tests.AssertSignature{
+		ret = "void",
+		func = "SetWater",
+		params = { self, loc, water, sink },
+		{ "userdata|GameBoard&", "userdata|Point", "boolean|bool", "boolean|bool" },
+		{ "userdata|GameBoard&", "userdata|Point", "boolean|bool" }
+	}
+	
+	if water then
+		if sink then
+			local d = SpaceDamage(loc)
+			d.iTerrain = TERRAIN_LAVA
+			self:DamageSpace(d)
+		else
+			self:SetTerrain(loc, TERRAIN_WATER)
+		end
+		
+		self:SetLava(loc, false)
+		self:SetAcid(loc, false)
+	elseif self:GetTerrain(loc) == TERRAIN_WATER then
+		self:SetTerrain(loc, TERRAIN_ROAD)
+	end
+end
```

Comment:
ℹ The sinking animation looks kind of awkward, since the tile changes to water first, and then starts sinking. Probably not possible to fix?

Reply:
The function is quite hacky. I have found addresses that could do the sinking without swapping terrain like this. There are a lot of moving parts involved though, so it would have to be at some point in the future.
If this function is not good enough at present, I can remove it.

Reply:
I have looked at this again, and I actually cannot see what you mean is the awkward part. It functions exactly like the acid barrels do, but without the splash. For a mod that wants to have a sinking tile, I think this function should be sufficient to work with.

Reply:
I see, I guess I never paid too much attention to it /shrug